tradeoffs.Rmd

```diff
@@ -0,0 +1,172 @@
+The doi.org system lets institutions register such (persistent) DOIs, and update (non-persistent) URLs where full texsts can currently be found.
+It might be helpful to query this API from R.
+
+Happily, the doi.org API is unauthenticated, highly performant and gives simple responses.
+For any given `GET` request, there are only these outcomes:
+
+1. A DOI cannot be found DOI.org (HTTP status code 402).
+2. A DOI can be found (HTTP status code 200).
+    a. The queried value can be found (API response code 1).
+    b. The queried value cannot be found (API response code 200).
+       For example, a DOI is registered but has no URL to resolve to.
+       Rare but theoretically possible.
+3. Something unspecified went wrong on the server end (HTTP status code 500).
+
+Omitting some bells and whistles from biblids, we can minimally wrap the API thus:
+
+```{r minimal-wrapper}
+# unexported helper
+# this is not tested, b/c the implementation could change
+verb_doi_handle <- function(doi, ...) {
+  httr::VERB(
+    url = "https://doi.org/",
+    # better to use curl::curl_escape() here
+    path = paste0("api/handles/", doi),
+    ...
+  )
+}
+
+# exported function
+# this is tested
+get_doi_handle <- function(doi, ...) {
+  resp <- verb_doi_handle(doi = doi, verb = "GET", ...)
+  httr::stop_for_status(resp) # covers outcomes 1 *and* 3 from above list
+  if (httr::content(resp)$responseCode == 200) {
+    # covers outcome 2b
+    stop("The handle exists, but the queried value(s) could not be found.")
+  }
+  # this must be outcome 2a
+  resp
+}
+
+# still missing: a way to disambiguate between outcomes 1 and 3
+# exported function, this is tested
+is_doi_found <- function(doi) {
+  # HEAD is more idiomatic, and marginally more efficient
+  resp <- verb_doi_handle(doi = doi, verb = "HEAD")
+  if (httr::status_code(resp) == 200) return(TRUE)
+  if (httr::status_code(resp) == 404) return(FALSE)
+  httr::stop_for_status(resp)
+}
+```
+
+A minimal interface between exported functions and the API is desirable, because it simplifies the below tests.
+`get_doi_handle()` would be the absolute minimum, but it alone cannot distinguish between API failure modes 1 from 3:
+`httr::stop_for_status()` could be thrown either because a DOI does not exist or because of some other HTTP error code.
+The differences between 1 and 3 are important for users and the downstream logic in the biblids package.
+We thus need an additional function at this interface, `is_doi_found()`, which checks explicity for failure mode 1.
+
+We can test most API failure modes of `get_doi_handle()` and `is_doi_found()` thus:
+
+```{r test-minimal-wrapper}
+library(testthat)
+skip_if_offline()
+# outcome 1
+test_that("Non-existent DOIs error out", {
+  expect_error(get_doi_handle("10.1000/I_DO_NOT_EXIST"))
+  expect_false(is_doi_found("10.1000/I_DO_NOT_EXIST"))
+})
+# outcome 2a
+test_that("Existing DOIs are found", {
+  testthat::local_edition(3)
```

Comment:
I'd argue this adds to much complexity that's unrelated to the matter of the chapter?

Reply:
yes indeed!